Add a warning when collecting 'async def' functions

As discussed in #2175, while it does not seem possible at the moment to support `asyncio` functions out of the box because it would be implemented in detriment of other async frameworks, we should at the very least emit a warning when collecting awaitable functions.

I tried both the approach of the original PR (#2175) and plugging in the checking if function is a coroutine [here](https://github.com/pytest-dev/pytest/blob/master/_pytest/python.py#L179) but no matter what I seem to do I get this stacktrace:

```
$ pytest testing/python/collect.py
====================================================================================== test session starts ======================================================================================
platform darwin -- Python 3.5.3, pytest-3.0.8.dev, py-1.4.33, pluggy-0.4.0
rootdir: /Users/pawelad/Dev/Code/Forks/pytest, inifile: tox.ini
plugins: hypothesis-3.6.1
collected 77 items 

testing/python/collect.py .........................F...................................................

=========================================================================================== FAILURES ============================================================================================
___________________________________________________________________________ TestFunction.test_async_function_warning ____________________________________________________________________________

self = <collect.TestFunction object at 0x10c795ef0>
testdir = <Testdir local('/private/var/folders/01/qtz5hrt13cxgy__flq7zjgb00000gn/T/pytest-of-pawelad/pytest-40/testdir/test_async_function_warning0')>

    @pytest.mark.skipif(sys.version_info < (3, 4),
                        reason='async features available in Python 3.4+')
    def test_async_function_warning(self, testdir):
        with pytest.warns(RuntimeWarning) as record:
            testdir.makepyfile("""
                    import asyncio
                    import types
                    @types.coroutine
                    def test_coroutine():
                        yield from asyncio.sleep(1)
                """)
    
>           reprec = testdir.inline_run()
E           Failed: DID NOT WARN. No warnings of type (<class 'RuntimeWarning'>,) was emitted. The list of emitted warnings is: [].

/Users/pawelad/Dev/Code/Forks/pytest/testing/python/collect.py:403: Failed
------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------
============================= test session starts ==============================
platform darwin -- Python 3.5.3, pytest-3.0.8.dev, py-1.4.33, pluggy-0.4.0
rootdir: /private/var/folders/01/qtz5hrt13cxgy__flq7zjgb00000gn/T/pytest-of-pawelad/pytest-40/testdir/test_async_function_warning0, inifile:
plugins: hypothesis-3.6.1
collected 0 items / 1 errors

==================================== ERRORS ====================================
_______________ ERROR collecting test_async_function_warning.py ________________

self = <CallInfo when='memocollect' exception: <function test_coroutine at 0x10c655d08> yielded non callable test <Future pending>>
func = <bound method Collector._memocollect of <Generator 'test_coroutine'>>
when = 'memocollect'

    def __init__(self, func, when):
        #: context of invocation: one of "setup", "call",
        #: "teardown", "memocollect"
        self.when = when
        self.start = time()
        try:
>           self.result = func()

/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/runner.py:163: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/main.py:476: in _memocollect
    return self._memoizedcall('_collected', lambda: list(self.collect()))
/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/main.py:347: in _memoizedcall
    res = function()
/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/main.py:476: in <lambda>
    return self._memoizedcall('_collected', lambda: list(self.collect()))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Generator 'test_coroutine'>

    def collect(self):
        # test generators are seen as collectors but they also
        # invoke setup/teardown on popular request
        # (induced by the common "test_*" naming shared with normal tests)
        from _pytest import deprecated
        self.session._setupstate.prepare(self)
        # see FunctionMixin.setup and test_setupstate_is_preserved_134
        self._preservedparent = self.parent.obj
        l = []
        seen = {}
        for i, x in enumerate(self.obj()):
            name, call, args = self.getcallargs(x)
            if not callable(call):
>               raise TypeError("%r yielded non callable test %r" %(self.obj, call,))
E               TypeError: <function test_coroutine at 0x10c655d08> yielded non callable test <Future pending>

/Users/pawelad/.pyenv/versions/3.5.3/envs/pytest/lib/python3.5/site-packages/_pytest/python.py:628: TypeError
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.28 seconds ============================
============================================================================== 1 failed, 76 passed in 6.47 seconds ==============================================================================
```

My test:
```python
    @pytest.mark.skipif(sys.version_info < (3, 4),
                        reason='async features available in Python 3.4+')
    def test_async_function_warning(self, testdir):
        with pytest.warns(RuntimeWarning) as record:
            testdir.makepyfile("""
                import asyncio
                import types
                @types.coroutine
                def test_coroutine():
                    yield from asyncio.sleep(1)
            """)

            reprec = testdir.inline_run()
            # reprec.assertoutcome(skipped=1)

        assert len(record) == 1
        assert record[0].message.args[0] == (
            "pytest doesn't currently support coroutines "
            "testing out of the box"
        )
```

Any idea what could be the reason?

--- a/changelog/2224.feature.rst
+++ b/changelog/2224.feature.rst
@@ -1,0 +1,4 @@
+``async`` test functions are skipped and a warning is emitted when a suitable
+async plugin is not installed (such as ``pytest-asyncio`` or ``pytest-trio``).
+
+Previously ``async`` functions would not execute at all but still be marked as "passed".

--- a/src/_pytest/python.py
+++ b/src/_pytest/python.py
@@ -43,6 +43,7 @@ from _pytest.mark import MARK_GEN
 from _pytest.mark.structures import get_unpacked_marks
 from _pytest.mark.structures import normalize_mark_list
 from _pytest.outcomes import fail
+from _pytest.outcomes import skip
 from _pytest.pathlib import parts
 from _pytest.warning_types import PytestWarning
 
@@ -156,6 +157,15 @@ def pytest_configure(config):
 @hookimpl(trylast=True)
 def pytest_pyfunc_call(pyfuncitem):
     testfunction = pyfuncitem.obj
+    iscoroutinefunction = getattr(inspect, "iscoroutinefunction", None)
+    if iscoroutinefunction is not None and iscoroutinefunction(testfunction):
+        msg = "Coroutine functions are not natively supported and have been skipped.\n"
+        msg += "You need to install a suitable plugin for your async framework, for example:\n"
+        msg += "  - pytest-asyncio\n"
+        msg += "  - pytest-trio\n"
+        msg += "  - pytest-tornasync"
+        warnings.warn(PytestWarning(msg.format(pyfuncitem.nodeid)))
+        skip(msg="coroutine function and no async plugin installed (see warnings)")
     funcargs = pyfuncitem.funcargs
     testargs = {arg: funcargs[arg] for arg in pyfuncitem._fixtureinfo.argnames}
     testfunction(**testargs)

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -1179,3 +1179,31 @@ def test_fixture_mock_integration(testdir):
 def test_usage_error_code(testdir):
     result = testdir.runpytest("-unknown-option-")
     assert result.ret == EXIT_USAGEERROR
+
+
+@pytest.mark.skipif(
+    sys.version_info[:2] < (3, 5), reason="async def syntax python 3.5+ only"
+)
+@pytest.mark.filterwarnings("default")
+def test_warn_on_async_function(testdir):
+    testdir.makepyfile(
+        test_async="""
+        async def test_1():
+            pass
+        async def test_2():
+            pass
+    """
+    )
+    result = testdir.runpytest()
+    result.stdout.fnmatch_lines(
+        [
+            "test_async.py::test_1",
+            "test_async.py::test_2",
+            "*Coroutine functions are not natively supported*",
+            "*2 skipped, 2 warnings in*",
+        ]
+    )
+    # ensure our warning message appears only once
+    assert (
+        result.stdout.str().count("Coroutine functions are not natively supported") == 1
+    )

